projects/ad9656_fmc/src/app/ad9656_fmc.c

```diff
@@ -0,0 +1,530 @@
+	buf[1] = reg_addr & 0xFF;
+	buf[2] = 0x00;
+
+	ret = spi_write_and_read(dev->spi_desc, buf, 3);
+	
+	if (ret < 0)
+		return ret;
+
+	*reg_data = buf[2];
+
+	return ret;
+}
+
+/**
+ * @brief Write to the ad9508 that is conected to the SPI
+ * @param dev - The device handler for the ad9508 chip
+ * @param reg_addr - Address of the internal register of the ad9508 chip
+ * @param reg_data - Value to be written to the register
+ * @return SUCCESS if the value was written successfully, FAILURE otherwise
+ */
+int32_t ad9508_spi_write(struct ad9508_dev *dev,
+			 uint16_t reg_addr,
+			 uint8_t reg_data)
+{
+	uint8_t buf[3];
+
+	// the MSB of byte 0 indicates a r/w operation, following 7 bits are the
+	// bits 14-8 of the address of the register that is accessed. Byte 1
+	// contains the bits 7-0 of the address of the register. Byte 2 contains
+	// the data to be written.
+	buf[0] = reg_addr >> 8;
+	buf[1] = reg_addr & 0xFF;
+	buf[2] = reg_data;
+
+	return spi_write_and_read(dev->spi_desc, buf, 3);
+}
+
+/**
+ * @brief Write to the ad9553 that is conected to the SPI
+ * @param dev - The device handler for the ad9553 chip
+ * @param reg_addr - Address of the internal register of the ad9553 chip
+ * @param reg_data - Value to be written to the register
+ * @return SUCCESS if the value was written successfully, FAILURE otherwise
+ */
+int32_t ad9553_spi_write(struct ad9553_dev *dev,
+			 uint16_t reg_addr,
+			 uint8_t reg_data)
+{
+	uint8_t buf[3];
+
+	// the MSB of byte 0 indicates a r/w operation, following 7 bits are the 
+	// bits 14-8 of the address of the register that is accessed. Byte 1 
+	// contains the bits 7-0 of the address of the register. Byte 2 contains
+	// the data to be written.
+	buf[0] = reg_addr >> 8;
+	buf[1] = reg_addr & 0xFF;
+	buf[2] = reg_data;
+
+	return spi_write_and_read(dev->spi_desc, buf, 3);
+}
+
+/**
+ * @brief Setup the working parameters of the ad9508 chip
+ * @param device - The device handler of the ad9508 chip
+ * @param init_param - Values for the working parameters of ad9508
+ * @return SUCCESS if device is ready for use, FAILURE otherwise
+ */
+int ad9508_setup(struct ad9508_dev **device,
+	     const struct ad9508_init_param *init_param)
+{
+	int32_t ret;
+	
+	struct ad9508_dev *dev;	
+	uint8_t reg_data;
+
+	dev = (struct ad9508_dev *)malloc(sizeof(*dev));
+	if (!dev)
+		return FAILURE;
+	
+	/* SPI */
+	ret = spi_init(&dev->spi_desc, &init_param->spi_init);
+	if (ret != SUCCESS)
+		return ret;
+
+	// reset 
+	ad9508_spi_write(dev, 0x00, 0x24);
+
+	// read family part id: 0x0C contains the least significant byte,
+	//                      0x0D contains the most significant byte
+	ad9508_spi_read(dev, 0x0C, &reg_data);
+	if (reg_data != 0x05)
+		return FAILURE;
+	ad9508_spi_read(dev, 0x0D, &reg_data);
+	if (reg_data != 0x00)
+		return FAILURE;
+	
+	// configure 9508 to pass the 125MHz input clock unmodified, so divider = 1
+	// no phase offset
+	ad9508_spi_write(dev, 0x1B, 0x00); // divide ratio[7:0]
```

Comment:
Would be nice to use macro defines for the Reg addresses.

Reply:
This statement is valid for the entire commit.

Reply:
Changed hardcoded register numbers to defined constants.